### DISCUSSION TITLE
Partial Traces of Operators in Constrained Hilbert Spaces

### DISCUSSION THREAD

**Discussion by connor-powers** [QuSpin/QuSpin#627] [General]
Hello! I'm defining a user_basis for a constrained Hilbert space, in that I'm defining a pre-check of the states that removes various disallowed states from the total Hilbert space. I'd like to perform a partial trace on an operator I construct that acts on this total constrained Hilbert space. Specifically, this operator is $e^{-tH}$ where the Hamiltonian H is constructed in QuSpin in the user_basis I've defined with the constraint. I saw the note under the "partial_trace" noting that the partial trace function can be applied to operators/observables defined by the input state, but I wasn't confident that I understood what this meant. Is this partial trace operation only supported for matrices which are explicitly constructed out of a state, e.g. density matrices, or can I apply this partial trace to a generic operator or observable that acts on the Hilbert space defined by my user_basis? 

If the latter, if I'm not further reducing the size of my Hilbert space (other than imposing the constraint) via working in a particular symmetry sector, is there nothing I have to do with regards to the caveat listed under the partial_trace() documentation? If it's helpful, I'm currently working with a machine that has QuSpin 0.3.6 installed, but I could probably get that updated if that would help. Thanks!

**Comment by mgbukov** [comment]
That's an interesting question; sometime ago I wanted to compute the
partial trace over a subsystem (say B) of an observable O, i.e. tr_B(O). I
think `basis.partial_trace()' should do the job in this case, even when
basis is a `user_basis' object.

I'm not sure if `partial_trace' will accept a `hamiltonian' object, though;
if it doesn't try to cast your operator as a sparse (and ultimately dense)
matrix before feeding it into `partial_trace`.

**In any case, I strongly urge you to test carefully the result using the
smallest system size available, to see if it does the job correctly.**

> Is this partial trace operation only supported for matrices which are
> explicitly constructed out of a state, e.g. density matrices, or can I
> apply this partial trace to a generic operator or observable that acts on
> the Hilbert space defined by my user_basis?
>
If it works properly, it should work for generic observables defined using
`user_basis`.

> If the latter, if I'm not further reducing the size of my Hilbert space
> (other than imposing the constraint) via working in a particular symmetry
> sector, is there nothing I have to do with regards to the caveat listed
> under the partial_trace() documentation?
>
If you're not using symmetries, that should not be an issue I guess.

> If it's helpful, I'm currently working with a machine that has QuSpin
> 0.3.6 installed, but I could probably get that updated if that would help.
> Thanks!
>
It's always better to use the most recent version; in 0.3.7 we fixed some
bugs, but I don't think they affect `partial_trace'.



> —
> Reply to this email directly, view it on GitHub
> <https://github.com/QuSpin/QuSpin/discussions/627>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADHSUBTD4XOBDTSB2CDN5B3XIDJZRANCNFSM6AAAAAAYQN42B4>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>


-- 
Marin Bukov, PhD
Research Group Leader
Max Planck Institute for the Physics of Complex Systems
Noethnitzer Str. 38
Dresden 01187, Germany

url: https://www.pks.mpg.de/nqd
<http://quantum-dynamics.phys.uni-sofia.bg/>

**Reply by connor-powers** [reply]
Okay, thank you for your reply!

**Reply by devashishtiwa** [reply]
Hi. 

Ma'am, while importing the partial trace from basis, its not being possible actually, it raises error. 
---------------------------------------------------------------------------
ImportError                               Traceback (most recent call last)
Cell In[6], [line 2](vscode-notebook-cell:?execution_count=6&line=2)
      [1](vscode-notebook-cell:?execution_count=6&line=1) from quspin import basis
----> [2](vscode-notebook-cell:?execution_count=6&line=2) from quspin.basis import partial_trace

ImportError: cannot import name 'partial_trace' from 'quspin.basis' (d:\conda_envs\quspin_env\lib\site-packages\quspin\basis\__init__.py)

this is what it says, i really dont understand why? 
I just wanted to compute to compute the partial trace of my own choice, like any spin, i like, i want to trace that out. 
as you said, similar to this, "I wanted to compute the
partial trace over a subsystem (say B) of an observable O, i.e. tr_B(O). I
think `basis.partial_trace()' should do the job in this case, even when
basis is a `user_basis' object.
"

**Comment by mgbukov** [comment]
partial_trace is not a function that you import; it's a method of the basis
class ;)
Marín
